### PR TITLE
Creating rust dependencies tree explorer

### PR DESCRIPTION
**Pull request by bruno-ortiz**
Hello! 

I tried to implement a tree view that shows the dependencies of a project.

It allows to see all dependencies to the project and it uses `cargo tree` for it. Also it allows to click and open the files, the viewtree tries its best to follow the openned file in the editor.

Here is an example:
image

Any feedback is welcome since i have basically no professional experience with TS.

### CODE REVIEW

**Comment by bruno-ortiz** [comment]
Hey guys! I don't want to be a annoying or anything like that, but as this is my first PR here, do i need to do something else for the PR to be reviewed?

**Review by Veykril — COMMENTED** [review]
I don't know my way around the client code too much, and I don't think any of the active people do either but overall this seems good to me.

Though I have two questions, how often does this invoke cargo (as in when) and we should ideally not have this code show error popups I think, if this code errors but the rest of r-a works fine I wouldn't personally want to bother users with it. So ideally errors should only be logged, and preferably in a way that is not too spammy.

**Comment by bjorn3** [comment]
Instead of using `cargo tree` should this use the dependency graph of rust-analyzer? That will also make it work when using `rust-project.json` instead of cargo and when using `rust-analyzer.linkedProjects`.

File: editors/code/src/ctx.ts [hunk 1]
```diff
@@ -42,7 +45,24 @@ export class Ctx {
...
+        const res = new Ctx(config, extCtx, client, serverPath, statusBar, dependenciesProvider, treeView);
+        res.pushCleanup(treeView);
+
+        vscode.window.onDidChangeActiveTextEditor(e => {
+            if (e && isRustEditor(e)) {
+                execRevealDependency(e).catch(reason => {
+                    void vscode.window.showErrorMessage(`Dependency error: ${reason}`);
+                });
+            }
+        });
```
[the remark this answers is not in the document]
Comment (bruno-ortiz) [reply editors/code/src/ctx.ts:65]: @Veykril Thanks for the review, i'll try to answer your questions:
1- It calls cargo tree is invoked every time the editor opens and after that for every "ok" that it receives from the server: The code is this:
```rust
 switch (status.health) {
            case "ok":
                this.statusBar.color = undefined;
                this.dependencies.refresh(); // it tells the editor that something changed in the server and it should update the dependency tree
                break;
            case "warning":
                this.statusBar.tooltip += "\nClick to reload.";
 ....
```
2- I agree that we should not show the popup errors. I can change that, what are the best practices to logging in the editor?
3- This is executed every time, so the dependency tree can "follow" the user. So every time the user changes the active editor, the dependency tree "follows" the user and reveals the file in it.

**Comment by bruno-ortiz** [comment]
@bjorn3 Can you please send me a reference or any example on how i can use the dependency graph?  And is this blocking, can we make this change afterwards?

Thanks for the comments

File: editors/code/src/ctx.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (bruno-ortiz) [reply editors/code/src/ctx.ts:65]: @lnicola it does not run cargo tree every time the editor changes, it runs when it receives a "status change" notification from the server. It checks that the status is "ok" and tries to reload the dependencies. 
But as pointed in the other comments ill try to look into the dependency graph feature.

File: editors/code/src/commands.ts [hunk 2]
```diff
@@ -246,6 +247,44 @@ export function openCargoToml(ctx: Ctx): Cmd {
...
+            await vscode.window.showTextDocument(uri);
+        } catch (err) {
+            await vscode.window.showErrorMessage(err.message);
+        }
+    };
+}
+
+export function revealDependency(ctx: Ctx): Cmd {
+    return async (editor: RustEditor) => {
+        const rootPath = vscode.workspace.workspaceFolders![0].uri.fsPath;
```
[the remark this answers is not in the document]
Comment (bruno-ortiz) [reply editors/code/src/commands.ts:262]: I'm not too familiar with VSCode API, i'll look into that.

File: editors/code/src/dependencies_provider.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    private async getRootDependencies(): Promise<Dependency[]> {
+        const registryDir = fspath.join(os.homedir(), '.cargo', 'registry', 'src');
+        const basePath = fspath.join(registryDir, fs.readdirSync(registryDir)[0]);
+        const deps = await this.getDepsInCartoTree(basePath);
+        const stdlib = await this.getStdLib();
+        this.dependenciesMap[stdlib.dependencyPath.toLowerCase()] = stdlib;
+        return [stdlib].concat(deps);
+    }
+
+    private async getStdLib(): Promise<Dependency> {
```
[the remark this answers is not in the document]
Comment (bruno-ortiz) [reply editors/code/src/dependencies_provider.ts:87]: Sure, can you please point me out on where can i find this logic in the server? ty

[the remark this answers is not in the document]
Comment (lnicola) [reply editors/code/src/dependencies_provider.ts:87]: There's https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide/src/view_crate_graph.rs for "View crate graph" and https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/project_model/src/workspace.rs#L140 takes care of loading a workspace. The sysroot is detected by `Sysroot::load` via `Sysroot::discover`.

**Comment by Veykril** [comment]
Ye, the server notifying the client here will have to be done as an lsp extension, a good inspiration for that would probably be the ServerStatus notification, as well as the view dependency graph as noted earlier.

File: editors/code/src/dependencies_provider.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nemethf) [reply editors/code/src/dependencies_provider.ts:87]: > This duplicates the logic in the server. I think we should do as much as possible there instead of in the client code.

I'm new here, but I disagree.  The server-side implementation takes some of the burden off of the developers of non-VSCode clients.

[the remark this answers is not in the document]
Comment (nemethf) [reply editors/code/src/dependencies_provider.ts:87]: I as well think it should be on the server side.  I'm sorry if I misunderstood something or if I wasn't clear.  Thanks.

**Comment by Veykril** [comment]
Went ahead and rebased this, this still needs to use the server for querying, I added a small commit implementing the scaffolding part for the lsp extension for now.

File: crates/ide/src/fetch_crates.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        .collect()
+}
+
+fn crate_name(data: &ide_db::base_db::CrateData) -> String {
+    data.display_name
+        .clone()
+        .map(|it| it.canonical_name().to_owned())
+        .unwrap_or("unknown".to_string())
+}
+
```
Comment (bruno-ortiz) [inline crates/ide/src/fetch_crates.rs:35]: I feel that the function below is SUPER hacky, but I couldn't figure another way around it.

**Comment by bruno-ortiz** [comment]
@Veykril sorry for the long delay, I got swamped at work.

I implemented the server query, feel free to review the code again..

File: crates/ide/src/fetch_crates.rs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+                        }
+                    }
+                    None => break,
+                }
+                root_path = vfs_path;
+            }
+            crate_path
+        }
+        None => "".to_owned(),
+    }
```
Comment (bruno-ortiz) [inline crates/ide/src/fetch_crates.rs:64]: I feel that this function is SUPER hacky, but I couldn't figure out another way around it.

File: docs/dev/lsp-extensions.md [hunk 6]
```diff
@@ -839,3 +839,26 @@ export interface Diagnostic {
...
+export interface FetchDependencyGraphParams {}
+```
+
+**Response:** 
+```typescript
+export interface FetchDependencyGraphResult {
+    crates: {
+        name: string;
+        version: string;
+        path: string;
```
Comment (nemethf) [inline docs/dev/lsp-extensions.md:860]: Wouldn't it be more general (and more in line with the language server protocol) to use `textDocument: TextDocumentIdentifier` here instead of the current `path: string`?

And as a side note: why this is called a graph/tree, when this is just a list?  It is not possible to construct a graph from the response.  Ie.: visualize the the dependencies of a dependency.

Comment (bruno-ortiz) [reply docs/dev/lsp-extensions.md:860]: @nemethf you're right about the naming, I changed from "graph" to "list". The tree is because the crates are displayed in a "ViewTree".
About the `TextDocumentIdentifier `, I tried using it, but it just made the code more complicated, because to reveal the items in the viewTree I need to work with the filesystem path, and the `TextDocumentIdentifier` inputs the "scheme" in the path.

File: editors/code/src/commands.ts [hunk 7]
```diff
@@ -260,6 +261,33 @@ export function openCargoToml(ctx: CtxInit): Cmd {
     };
 }
 
+export function revealDependency(ctx: CtxInit): Cmd {
+    return async (editor: RustEditor) => {
+        const rootPath = vscode.workspace.workspaceFolders![0].uri.fsPath;
+        const documentPath = editor.document.uri.fsPath;
+        if (documentPath.startsWith(rootPath)) return;
```
Comment (davidbarsky) [inline editors/code/src/commands.ts:268]: I'm not entirely sure why you'd want to do this. There's already non-workspace filtering done on the server; this just adds an additional check.

(This also breaks for monorepos where users open their editor at the monorepo root. Non-workspace crates are often stored in `third-party/` but share the workspace root.)

File: editors/code/src/commands.ts [hunk 8]
```diff
@@ -260,6 +261,33 @@ export function openCargoToml(ctx: CtxInit): Cmd {
...
+        } else {
+            let documentPath = editor.document.uri.fsPath;
+            const parentChain: DependencyId[] = [{ id: documentPath.toLowerCase() }];
+            do {
+                documentPath = path.dirname(documentPath);
+                parentChain.push({ id: documentPath.toLowerCase() });
+            } while (!ctx.dependencies?.contains(documentPath));
+            parentChain.reverse();
+            for (const idx in parentChain) {
+                await ctx.treeView?.reveal(parentChain[idx], { select: true, expand: true });
```
Comment (davidbarsky) [inline editors/code/src/commands.ts:281]: do you mind adding a comment explaining why this bit is necessary? is it to show a dependency of a dependency?

File: crates/ide/src/fetch_crates.rs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+) -> Option<String> {
+    let source_root_id = db.file_source_root(data.root_file_id);
+    let source_root = db.source_root(source_root_id);
+    let source_root_path = source_root.path_for_file(&data.root_file_id);
+    source_root_path.cloned().and_then(|mut root_path| {
+        let mut crate_path = None;
+        while let Some(vfs_path) = root_path.parent() {
+            match vfs_path.name_and_extension() {
+                Some((name, _)) => {
+                    if name.starts_with(crate_name) {
```
Comment (davidbarsky) [inline crates/ide/src/fetch_crates.rs:60]: the file path component here isn't necessarily guaranteed to start with crate name. i mentioned this in another point, but I think getting the full URI of the crate would probably make some other parts in this PR easier to implement.

File: crates/ide/src/fetch_crates.rs (hunk 5 above)
Comment (bruno-ortiz) [reply crates/ide/src/fetch_crates.rs]: @davidbarsky this function tries to find a crate root dir in the filesystem, through the CrateGraph I can get the root file of a crate, for example, tokio's root file is `/home/<user>/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/lib.rs`. But I don't need the root file i need the root dir of a crate, so I need to convert this file to `/home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0`.
If there is a cleaner way of doing it, I'll gladly change the implementation.

File: editors/code/src/commands.ts (hunk 7 above)
Comment (bruno-ortiz) [reply editors/code/src/commands.ts:268]: The purpose of the function revealDependency is to make the TreeView "reveal" the opened file in the editor. Example:
image

But if I where to open a file that belongs to the local project, I dont want to trigger any "revealing"

File: editors/code/src/commands.ts (hunk 8 above)
Comment (bruno-ortiz) [reply editors/code/src/commands.ts:281]: I'll add it.
Its because the TreeView is Lazy, so at first it only has the root dependencies: Example:
- core
- alloc
- std

if I want to reveal alloc/src/str.rs, I have to:
1. reveal every children of alloc
- core
- alloc
&emsp;|-beches
&emsp;|-src
&emsp;|- ...
- std
2. reveal every children of src:
- core
- alloc
&emsp;|-beches
&emsp;|-src
&emsp;&emsp;| - lib.rs
&emsp;&emsp;| - str.rs    <------- FOUND IT!
&emsp;&emsp;| - ...
&emsp;|- ...
- std

So basically, i have to calculate the parents of the file that I want to reveal until i find the root crate path.

File: crates/ide/src/fetch_crates.rs (hunk 9 above)
Comment (bruno-ortiz) [reply crates/ide/src/fetch_crates.rs:60]: I agree, the problem here is that i could no find the full URI of the crate anywhere, maybe someone with more experience in the RA API could help me.

File: docs/dev/lsp-extensions.md (hunk 6 above)
Comment (nemethf) [reply docs/dev/lsp-extensions.md:860]: > The tree is because the crates are displayed in a "ViewTree". 

Thank you.  I have no objections against it.  However, it is a bit strange that a client-side implementation detail determines an abstract name in protocol specification.  I can imagine some clients using this extension for something else, for example, to provide a "jump to dependency" functionality that implements an incremental search on the dependencies without showing them beforehand. 

Also this fetchDependencyGraph reminds me of the rust-analyzer/viewCrateGraph LSP extension.  In the long run, it might make sense to generalize both of them to send the same details.   fetchDependencyGraph could list the dependencies of a dependency, and viewCrateGraph could be extended to have a uri/path for each dependency.

**Comment by bruno-ortiz** [comment]
I converted this PR to draft because I detected some problems working with multi-workspaces. 
I'll try to fix it ASAP.

File: editors/code/src/commands.ts (hunk 7 above)
Comment (davidbarsky) [reply editors/code/src/commands.ts:268]: > The purpose of the function revealDependency is to make the TreeView "reveal" the opened file in the editor. 

Gotcha, thanks for explaining—I've built the branch locally and usage made a bit more sense.

> But if I where to open a file that belongs to the local project, I dont want to trigger any "revealing"

That makes sense and works for most default Cargo-based builds. However, using `if (documentPath.startsWith(rootPath)) return;` in situations where `cargo vendor` has been run _or_ when a user opens their VS Code instance at the root of the monorepo (where a `third-party/` folder exists and contains vendored Rust crates), this check will prevent this tree-based explorer from ever opening and it seems unfortunate that those users won't benefit from this feature[^1]. I don't have merge permissions on rust-analyzer, but I'd recommend removing this check, as `fetch_crates` in `crates/ide/src/fetch_crates.rs` already performs a similar check. I need to look over this PR a bit more, but if a user asks to reveal the dependency for a crate that is part of their editable workspace, then I would consider showing it. Alternatively, it might make sense to add another LSP extension to query the server itself to check whether a file is editable/part of the current editable workspace or not. 

[^1]: Personally, I'd love to build atop of this UI in order to allow users to incrementally add crates to their (editable) rust-analyzer workspace—it's a pretty workflow in monorepos.

Comment (bruno-ortiz) [reply editors/code/src/commands.ts]: @davidbarsky thanks for the reply, I just pushed a commit improving some pain points and hopefully giving it proper support to multiple workspace folders.
Could you point me at any github repos that make use of the features that you mentioned? I would love to test this cases because I suspect that they won't be a problem, because if they're already part of the project tree, the main treeView of vscode will already display those dependencies.

Comment (bruno-ortiz) [reply editors/code/src/commands.ts]: Just to add to my previous comment, I created a vendored project and everything seems fine.
image

**Comment by bruno-ortiz** [comment]
Marking this as ready to review again, I thought it would take longer to fix the multi-workspace issues.

File: editors/code/src/commands.ts (hunk 7 above)
Comment (davidbarsky) [reply editors/code/src/commands.ts]: > Could you point me at any github repos that make use of the features that you mentioned?

@bruno-ortiz I can't think of any project off-hand, but it would be possible to take, e.g., the rust-analyzer repository and run `cargo vendor` and you'll see a subset of the behavior I'm mentioning.

> I would love to test this cases because I suspect that they won't be a problem, because if they're already part of the project tree, the main treeView of vscode will already display those dependencies.

At the risk of being *that person*, but a smaller repository won't really demonstrate this issue—it only really becomes apparent once a monorepo has tens (or hundreds!) of millions of lines of code. While said dependencies _are_ technically part of the primary explorer view, it's often infeasible to see them at a glance because of the sheer number of folders present in the default file explorer.

My mental model of this feature—at least in monorepos—is that it is a subset of the monorepo (think of it as a materialized view...) that corresponds to the portions of monorepo most relevant to the code being worked on at the moment.

The reason that I'm so adamant about requesting this is that I didn't really _get_ the need/desire for it until I tried it. While it is technically possible to navigate through the primary explorer window to dependencies, it ends up taking closer to 30-60 seconds per dependency rather than something instantaneous.

**Comment by bruno-ortiz** [comment]
@davidbarsky @nemethf I've made some quality of life improvements and made some best effort attempts to support the every usecase possible. One pain point is:

I could not figure out if a crate is vendored or not. Since RA uses "cargo metadata" to fetch the info about the crates, it receives this info from cargo itself, so unless we are able to parse cargo's config(config.toml), I think we will not be able to get a hold of this information. This doesn't mean that it doesn't works for projects with vendored crates, the dependencies will be shown, but the automatic navigation will be turned off, since vscode itself will auto-navigate using the primary explorer.

Does anyone see any other issues that would make this PR unmergeable?

**Comment by nemethf** [comment]
> Does anyone see any other issues that would make this PR unmergeable?

I've only looked at lsp-extensions.md, and it is not a showstopper for me, but there is a bit of an inconsistency: the fetchDependencyList has FetchDependencyGraphParams and FetchDependencyGraphResult.  I think it should be FetchDependencyListParams and FetchDependencyListResult.

Thank you.

File: crates/base-db/src/input.rs [hunk 10]
```diff
@@ -304,6 +304,7 @@ pub struct CrateData {
     /// For purposes of analysis, crates are anonymous (only names in
     /// `Dependency` matters), this name should only be used for UI.
     pub display_name: Option<CrateDisplayName>,
+    pub crate_root_path: Option<AbsPathBuf>,
```
Comment (Veykril) [inline crates/base-db/src/input.rs:307]: We shouldn't store any paths like this in `CrateData`, instead we should resolve the `root_file_id` back to its path where we need it

File: crates/ide/src/fetch_crates.rs [hunk 11]
```diff
@@ -0,0 +1,45 @@
+use ide_db::{
+    base_db::{CrateOrigin, SourceDatabase},
+    FxIndexSet, RootDatabase,
+};
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct CrateInfo {
+    pub name: String,
+    pub version: String,
+    pub path: String,
```
Comment (Veykril) [inline crates/ide/src/fetch_crates.rs:10]: This should probably be a `FileId` instead then (given the previous comment),

File: crates/paths/src/lib.rs [hunk 12]
```diff
@@ -213,6 +213,13 @@ impl AbsPath {
     pub fn exists(&self) -> bool {
         self.0.exists()
     }
+
+    pub fn name_and_extension(&self) -> Option<(&str, Option<&str>)> {
+        Some((
+            self.file_stem()?.to_str()?,
+            self.extension().and_then(|extension| extension.to_str()),
+        ))
+    }
```
Comment (Veykril) [inline crates/paths/src/lib.rs:222]: This isn't a delegated method so we shouldn't put that into the `region` comment area, though assuming this is only in one place we might as well replace the function call with what we do here

File: crates/rust-analyzer/src/handlers.rs [hunk 13]
```diff
@@ -125,6 +128,19 @@ pub(crate) fn handle_memory_usage(state: &mut GlobalState, _: ()) -> Result<Stri
...
 
+pub(crate) fn fetch_dependency_list(
+    state: GlobalStateSnapshot,
+    _params: lsp_ext::FetchDependencyListParams,
+) -> Result<lsp_ext::FetchDependencyListResult> {
+    let crates = state.analysis.fetch_crates()?;
+    Ok(FetchDependencyListResult {
+        crates: crates
+            .into_iter()
+            .map(|it| CrateInfoResult { name: it.name, version: it.version, path: it.path })
```
Comment (Veykril) [inline crates/rust-analyzer/src/handlers.rs:139]: Here we need to resolve the file id to its path, or url even, with the `GlobalStateSnapshot`

File: crates/rust-analyzer/tests/slow-tests/tidy.rs [hunk 14]
```diff
@@ -45,7 +45,7 @@ fn check_lsp_extensions_docs() {
             sh.read_file(sourcegen::project_root().join("docs/dev/lsp-extensions.md")).unwrap();
         let text = lsp_extensions_md
             .lines()
-            .find_map(|line| line.strip_prefix("lsp_ext.rs hash:"))
+            .find_map(|line| dbg!(line.strip_prefix("lsp_ext.rs hash:")))
```
Comment (Veykril) [inline crates/rust-analyzer/tests/slow-tests/tidy.rs:48]: leftover dbg

File: editors/code/src/dependencies_provider.ts [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+                });
+                return resolve(files);
+            } else {
+                return resolve(this.getRootDependencies());
+            }
+        });
+    }
+
+    private async getRootDependencies(): Promise<Dependency[]> {
+        const dependenciesResult: FetchDependencyListResult = await this.ctx.client.sendRequest(
```
Comment (Veykril) [inline editors/code/src/dependencies_provider.ts:78]: when exactly is this being fired? This should ideally only trigger when something relevant to project changes

File: docs/dev/lsp-extensions.md (hunk 6 above)
Comment (Veykril) [reply docs/dev/lsp-extensions.md:860]: A `TextDocumentIdentifier` would indeed make more sense if possible

**Comment by Veykril** [comment]
I believe this will show duplicate entries for project that have multiple cargo workspaces/linked projects loaded. Not a huge problem but something we should be aware of (this would be fixed once we deduplicate crates in the crate graph).

File: crates/base-db/src/input.rs (hunk 10 above)
Comment (bruno-ortiz) [reply crates/base-db/src/input.rs:307]: Ok, I was unsure that this was the right aproach, but could you elaborate a little more on why this is bad?

File: crates/rust-analyzer/src/handlers.rs [hunk 16]
```diff
@@ -1916,3 +1939,24 @@ fn run_rustfmt(
...
+                        break;
+                    }
+                }
+                None => break,
+            }
+            root_path = path;
+        }
+        crate_path
+    })
+}
```
Comment (Veykril) [inline crates/rust-analyzer/src/handlers.rs:1962]: This at least deserves a comment explaining what it does as it is not clear at all. Though we could be smarter about this I think. 
We can iterate over all cargo workspaces we know, iterate over the pacakages there, then iterate over the targets that belong to those packages and check if the targets root file path is the same as the path we are looking for. We now have a package with a fitting root path, from there we can read out the cargo toml path of the package and go up to the dir.

(this is obviously quadratic so we probably want to do some upfront hashmap collecting once to reduce the iteration amount required.

File: crates/rust-analyzer/src/lsp_ext.rs [hunk 17]
```diff
@@ -27,6 +27,31 @@ pub struct AnalyzerStatusParams {
     pub text_document: Option<TextDocumentIdentifier>,
 }
 
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct CrateInfoResult {
+    pub name: Option<String>,
+    pub version: Option<String>,
+    pub path: String,
```
Comment (Veykril) [inline crates/rust-analyzer/src/lsp_ext.rs:35]: was changing this a to `TextDocumentIdentifier` still not possible?

**Comment by Veykril** [comment]
Also please remove the merge commits, we prefer to not introduce merge commits aside from the ones bors introduces.

File: crates/rust-analyzer/src/lsp_ext.rs (hunk 17 above)
Comment (bruno-ortiz) [reply crates/rust-analyzer/src/lsp_ext.rs:35]: I really don't think that this should be a TextDocumentIdentifier, because it represents the path to the crate's directory, not to a text document.

**Comment by bruno-ortiz** [comment]
@Veykril I added the requested comment and rebased the commits onto master branch.

File: crates/base-db/src/input.rs (hunk 10 above)
Comment (Veykril) [reply crates/base-db/src/input.rs]: We want as little code to know about paths as possible. Paths are tricky to handle correctly as different paths can describe the same file etc, FileId's are unambiguous. And in this case, we already store the file id, so storing the path as well is redundant.

File: crates/rust-analyzer/src/handlers.rs [hunk 18]
```diff
@@ -1910,3 +1932,34 @@ fn run_rustfmt(
...
+                    return Some(path);
+                }
+            } else {
+                break;
+            }
+            root_path = path;
+        }
+        None
+    })
+}
```
Comment (Veykril) [inline crates/rust-analyzer/src/handlers.rs:1965]: Okay, so this will not work too well with crates that are renamed in the cargo toml causing their folder name to differ from the crate name. I think a better approach here is to assume the classic library structure, go up two steps and check whether the `Cargo.toml` is where we expect it. That is check that `{parentPath}/../Cargo.toml` exists, then return `{parentPath}/../..`, so we assume `../crate/src/lib.rs` as the crate root (which is almost always the case for libraries)

**Comment by Veykril** [comment]
Two more things then I think its ready to go 🎉 
(though I'll wait with merging until next week just to have some nightly time for it)

File: crates/rust-analyzer/src/handlers.rs (hunk 18 above)
Comment (bruno-ortiz) [reply crates/rust-analyzer/src/handlers.rs]: nice, I changed the code meet your suggestion, it made the code simpler 👍

**Comment by bruno-ortiz** [comment]
@Veykril I made the requested fixes:

* changed the type of the returned path to be an Url;
* changed the crate_path function to lookup for the Cargo.toml dir.

**Comment by bruno-ortiz** [comment]
@Veykril I fixed the merging conflicts but I don't have merge permissions.
